scylla/src/transport/connection.rs

```diff
@@ -580,40 +588,54 @@ impl Connection {
             timestamp: batch.get_timestamp(),
         };
 
-        loop {
-            let query_response = self
-                .send_request(&batch_frame, true, batch.config.tracing)
-                .await?;
+        let query_response = self
+            .send_request(&batch_frame, true, batch.config.tracing)
+            .await?;
 
-            return match query_response.response {
-                Response::Error(err) => match err.error {
-                    DbError::Unprepared { statement_id } => {
-                        debug!("Connection::batch: got DbError::Unprepared - repreparing statement with id {:?}", statement_id);
-                        let prepared_statement = batch.statements.iter().find_map(|s| match s {
-                            BatchStatement::PreparedStatement(s) if *s.get_id() == statement_id => {
-                                Some(s)
-                            }
+        match query_response.response {
+            Response::Result(_) => Ok(BatchResult {
+                warnings: query_response.warnings,
+                tracing_id: query_response.tracing_id,
+            }),
+            Response::Error(err) => match err.error {
+                DbError::Unprepared { .. } => {
+                    debug!("Connection::batch: got DbError::Unprepared - repreparing all statements in the batch");
+
+                    let reprepares_futures = batch
+                        .statements
+                        .iter()
+                        .filter_map(|s| match s {
+                            BatchStatement::PreparedStatement(s) => Some(s),
                             _ => None,
+                        })
+                        .map(|prepared_statement| {
```

Comment:
It's probable that a single batch will have multiple instances of the same prepared query, only with different arguments - for example imagine 100 inserts to the same table. It would be good to deduplicate them before repreparing.

Reply:
I assume you mean deduplicating by `prepared_id`, am I right?

Reply:
Yup